# no-cross-layer-import

Prevents upper layers from importing lower layers in Feature-Sliced Design (FSD) architecture.

## ğŸ“– Rule Details

Feature-Sliced Design (FSD) architecture has a hierarchical structure where each layer can only import from layers below it.

**FSD Layer Order (top to bottom):**

```
app      (Top)
  â†“
pages
  â†“
widgets
  â†“
features
  â†“
entities
  â†“
shared   (Bottom)
```

**Allowed import directions:**
- âœ… Upper layer â†’ Lower layer (e.g., `pages` â†’ `widgets`)
- âœ… Same layer (e.g., `features/auth` â†’ `features/auth/ui`)
- âŒ Lower layer â†’ Upper layer (e.g., `entities` â†’ `features`)

## ğŸ”´ Examples of Incorrect Code

```javascript
// âŒ pages importing app (lower importing upper)
// File: src/pages/home/index.js
import { config } from '@/app/config';

// âŒ widgets importing pages
// File: src/widgets/header/Header.js
import { HomePage } from '@/pages/home';

// âŒ features importing widgets
// File: src/features/auth/index.js
import { Sidebar } from '@/widgets/sidebar';

// âŒ entities importing features
// File: src/entities/user/model.js
import { login } from '@/features/auth';

// âŒ shared importing entities
// File: src/shared/ui/Avatar.js
import { User } from '@/entities/user';

// âŒ require syntax is also checked
// File: src/features/auth/index.js
const Header = require('@/widgets/header');

// âŒ dynamic import is also checked
// File: src/widgets/header/index.js
const module = await import('@/pages/home');
```

## ğŸŸ¢ Examples of Correct Code

```javascript
// âœ… app importing pages (upper importing lower)
// File: src/app/App.js
import { MainPage } from '@/pages/main';

// âœ… pages importing widgets
// File: src/pages/home/index.js
import { Header } from '@/widgets/header';

// âœ… widgets importing features
// File: src/widgets/sidebar/Sidebar.js
import { LoginForm } from '@/features/auth';

// âœ… features importing entities
// File: src/features/profile/index.js
import { User } from '@/entities/user';

// âœ… entities importing shared
// File: src/entities/post/ui/PostCard.js
import { Button } from '@/shared/ui';

// âœ… Same layer imports
// File: src/entities/user/index.js
import { UserCard } from './UserCard';

// âœ… External package imports
// File: src/pages/home/index.js
import React from 'react';
import { useQuery } from 'react-query';
```

## âš™ï¸ Options

### `alias`

Specify the path alias prefix. Default is `"@"`.

```json
{
  "rules": {
    "@yh-kim/fsd/no-cross-layer-import": ["error", {
      "alias": "@"
    }]
  }
}
```

### `ignorePatterns`

Specify file patterns to ignore as an array of regular expressions.

```json
{
  "rules": {
    "@yh-kim/fsd/no-cross-layer-import": ["error", {
      "ignorePatterns": [
        "\\.test\\.",
        "\\.spec\\.",
        "/tests/",
        "/stories/"
      ]
    }]
  }
}
```

## ğŸ’¡ When to Use

This rule is useful when:

- Implementing Feature-Sliced Design architecture in your project
- You want to manage layer dependencies explicitly
- You want to prevent circular dependencies
- You want to enforce architectural structure in your codebase

## ğŸ”— Further Reading

- [Feature-Sliced Design Official Documentation](https://feature-sliced.design/)
- [FSD - Architectural Requirements](https://feature-sliced.design/docs/reference/layers)

## âš¡ Implementation Details

This rule checks the following import statements:

- ES6 `import` statements
- CommonJS `require()` calls
- Dynamic `import()` expressions

**Supported path formats:**
- Absolute alias (`@/entities/user`, `~/features/auth`)
- Relative paths (`../../entities/user`)

**Cases not checked:**
- External package imports (`react`, `lodash`, etc.)
- Files not in FSD layer directories
- Imports within the same layer
